### PR TITLE
Math Verification Suite + D7 boundary: implement the Math Correctness Charter (roadmap 0.11)

### PR DESCRIPTION
**Pull request by bberardi**
## Summary

Implements roadmap item **0.11 — the Math Verification Suite**, the final Phase 0 work item and the **gate before Phase 2** puts numbers in front of users. This makes the Math Correctness Charter (§4) executable: every verification layer, the 100% coverage gate, mutation testing, and the D7 core/UI boundary. One commit per subsection (the roadmap's suggested split, chunked into one PR).

Test suite grew **235 → 288**; `src/helpers/**` is at **100% line + branch + function** coverage; lint, format, and build green at every step.

| Commit | Subsection | What it does |
| --- | --- | --- |
| `dda3e2c` | **D7 boundary + hook move** | ESLint `no-restricted-imports` rules scope `src/helpers/**` and `src/models/**` to a framework-free core — no `react`, `react-dom`, `@mui/*`, `@emotion/*`, `react-number-format`, or `*.tsx` (UI component) imports. The lone offender, `useFieldTracking` (a React hook living in `helpers/`), moves to `src/hooks/`. The boundary is now a build failure, not a convention. |
| `9ee0e13` | **Verification suite + precision policy** | Charter layers 1–5: `math-reference.test.ts` (oracle tests vs Excel PMT/IPMT/CUMIPMT/FV and a hand-derived schedule, sources cited), `forecast-consistency.test.ts` (engine reproduces the schedule helpers to the cent), `math-properties.test.ts` (fast-check invariants), `math-edge-cases.test.ts` (leap days, month-end rollover, negative amortization, 600+-month float accumulation, …), and `PRECISION.md` (the one documented rounding/consistency policy). Adds `fast-check`. |
| `0b54022` | **Coverage gate** | v8 coverage scoped to `src/helpers/**` with a hard **100% line+branch threshold** wired into CI (`npm run test:coverage`). Closes the gaps with `defensive-guards.test.ts` rather than weakening the helpers; a few behavior-preserving cleanups remove genuinely dead branches. |
| `4c4931d` | **Mutation testing + docs** | `stryker.config.json` + a **weekly** `mutation.yml` workflow (off the per-PR path; uploads the HTML report). Updates `copilot-instructions.md` with the boundary, the Charter standard, and the new `src/hooks/` location. |

## Verification layers (Charter §4)

1. **Reference (oracle)** — ≥2 independent points per formula, each citing its spreadsheet/closed-form source.
2. **Cross-implementation consistency** — `forecastLoan` reproduces `generateAmortizationSchedule` month-for-month to the cent; `forecastInvestment` matches `generateInvestmentGrowth` to the cent at every compounding boundary.
3. **Property/invariant** (fast-check) — money conservation in cents, non-negative balances, monotonic extra payments/contributions, pointwise net-worth additivity, step-up monotonicity.
4. **Edge-case catalog** — leap-day anniversaries, month-end rollover, zero/extreme rates, one-month terms, negative amortization, horizon=today / mid-month / 50-year.
5. **Precision policy** — `PRECISION.md`: cents, round-half-up, where each module rounds; exact `toBe` where defined, every `toBeCloseTo` justified.

## A finding worth flagging

The cross-implementation tests surfaced a **genuine divergence**: with yearly **step-ups** enabled, the two investment engines attribute an on-anniversary contribution to different year numbers (an off-by-one). Without step-ups they are identical to the cent, so it is purely a step-up *timing* question. It is documented in `PRECISION.md` §4 and the consistency suite asserts only the provably-exact no-step-up cases. Per the Charter's process rule, reconciling it should ship as its own PR with a failing regression test landed before the fix — I did **not** silently change forecast behavior here.

## Notes for review

- No user-facing behavior change; version stays `0.7.0` (the Phase 0 target). The `investment-helpers` tweaks (dead-branch removal, hoisting a redundant `|| 0`) are behavior-preserving and covered by the existing suite.
- New dev deps only: `fast-check`, `@vitest/coverage-v8`, `@stryker-mutator/core` + `vitest-runner`. `npm ci` verified clean under both npm 10 and npm 11 (Node 24).
- The 100% bar is the math core only; UI keeps pragmatic targets (Phase 6), per the Charter.
- **With this merged, Phase 0 is complete** — Phase 1 (persistence, #20) and Phase 2 (charts, #18) are unblocked, and the chart numbers now sit on a verified engine.

https://claude.ai/code/session_013L3NNdNfJM14L9KAnKkG5x

---
_Generated by [Claude Code](https://claude.ai/code/session_013L3NNdNfJM14L9KAnKkG5x)_

### CODE REVIEW

File: src/helpers/PRECISION.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+### Known divergence (documented, not yet reconciled)
+
+With **yearly step-ups enabled**, the two investment engines assign a
+contribution that lands on an anniversary to **different year numbers** (an
+off-by-one in anniversary attribution: the period-indexed
+`generateInvestmentGrowth` first steps up one period later than the monthly-grid
+`forecastInvestment`). Without step-ups the engines are identical to the cent, so
+this is purely a step-up **timing** question, not a compounding error. It is
+called out here and in the consistency suite as a known gap; per the Charter's
+process rule, reconciling it ships as its own PR with a failing regression test
```
Comment (bberardi) [inline src/helpers/PRECISION.md]: **Capture this divergence as an executable, visible test — not just prose.**

This is the one genuinely actionable finding, and the PR handles it honestly. The step-up off-by-one is a real `src/helpers/**` correctness defect, and ROADMAP.md §4 (Math Correctness Charter, "Process rules") is explicit: *"every math bug found — by review, by a user, by chance — gets a failing regression test committed before the fix."* It's also exactly the class of bug PR #60's proposed §8.1 correctness backlog is trying to round up.

Right now the divergence lives only in this Markdown and as an *absence* of an assertion in `forecast-consistency.test.ts` — so a `npm test` run is green and silent about it. That's the failure mode the Charter exists to prevent (a "plausible chart on a subtly wrong formula", §4 intro). Two cheap ways to make it visible without changing forecast behavior in this PR:

1. Add the step-up case to `forecast-consistency.test.ts` as `it.fails(...)` (or `it.todo(...)`), so the divergence shows up in the runner and flips to a normal `it` the moment the fix PR lands — which is the "failing regression test committed before the fix" the Charter asks for.
2. Open a tracking issue and link it here, so it joins the §8.1 backlog rather than being deferred indefinitely.

Either is fine; the point is that the gap should be enforced by the suite, not by reviewers remembering this paragraph.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: stryker.config.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  "mutate": ["src/helpers/**/*.ts", "!src/helpers/**/*.test.ts"],
+  "reporters": ["html", "clear-text", "progress"],
+  "htmlReporter": {
+    "fileName": "reports/mutation/index.html"
+  },
+  "_thresholds_comment": "break is a conservative floor so the scheduled run reports a score rather than flaking red; the Charter target is zero surviving mutants, so ratchet break upward as survivors are triaged.",
+  "thresholds": {
+    "high": 90,
+    "low": 80,
+    "break": 70
```
Comment (bberardi) [inline stryker.config.json]: `break: 70` is a sensible "don't flake red" floor for a first scheduled run, and the comment above it is honest about it being a ratchet. One gap worth noting against ROADMAP.md §4: the Charter's mutation rule is *"surviving mutants triaged to zero or explicitly waived with a comment"* — but a weekly, non-blocking workflow with a 70% break threshold has **no mechanism that forces the triage to ever happen**. Nothing fails, nothing nags.

Suggest one of: (a) record the initial mutation score in the PR description / an issue so there's a baseline to ratchet from, and (b) commit to bumping `break` toward the actual score after the first run. Otherwise this layer is wired up but inert — coverage proves the lines ran, and per §4 mutation is the layer that proves the tests would *catch* a flipped sign. Not a blocker for this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/helpers/math-edge-cases.test.ts [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    // Jan 31 + 1 month overflows February and normalizes into March — a known
+    // Date quirk. Pinned so a future dayjs migration (Phase 6) is a deliberate
+    // behavior change, not a silent one.
+    const next = getNextCompoundingDate(
+      new Date(2021, 0, 31),
+      CompoundingFrequency.Monthly
+    );
+    expect(next.getMonth()).toBe(2); // March
+    expect(next).toEqual(new Date(2021, 2, 3));
+  });
```
Comment (bberardi) [inline src/helpers/math-edge-cases.test.ts:73]: Good call pinning the JS `Date` Jan-31 + 1-month → Mar 3 rollover — this is precisely the "month-end dates (Jan 31 + 1 month)" entry in ROADMAP.md §4's edge-case catalog, so the layer-4 box is genuinely ticked.

Two notes, neither blocking:
- This test *documents* the quirk rather than asserting correct calendar behavior — a contribution that should compound in February silently lands in March. The comment frames it as a deliberate pin for the Phase 6 dayjs migration, which is the right call given D7 says the core is "TypeScript + dayjs only" and this engine still uses raw `Date`. Worth a tracking issue so the dayjs migration explicitly revisits it rather than inheriting the quirk by default.
- Minor: `expect(next).toEqual(new Date(2021, 2, 3))` over-specifies the day (3) on top of the `.getMonth()` assertion. That's fine as a tight pin, just be aware it's asserting the *quirk's* exact output, so the dayjs migration will (correctly) have to change this line.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/helpers/investment-helpers.ts [hunk 4]
```diff
@@ -116,8 +116,10 @@ export const getInvestmentPeriods = (
     }
   }
 
-  // Return at least 1 period if the start date is today or in the past
-  return Math.max(periods, start <= end ? 1 : 0);
+  // Return at least 1 period: an investment whose start is today or in the
+  // past has elapsed at least one period. (end < start already returned 0
+  // above, so end >= start holds here.)
+  return Math.max(periods, 1);
```
Comment (bberardi) [inline src/helpers/investment-helpers.ts:122]: Confirmed behavior-preserving ✅ — the early `if (end < start) return 0;` at the top of the function guarantees `end >= start` here, so `start <= end ? 1 : 0` was always `1`. Collapsing it to `Math.max(periods, 1)` removes a genuinely dead branch (exactly the kind the §4 100% **branch** gate would otherwise be unable to cover without weakening the helper). The PR's framing of these `investment-helpers` tweaks as dead-branch removal is accurate.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by bberardi — COMMENTED** [review]
## Review: roadmap 0.11 — Math Verification Suite + D7 boundary

Reviewed against `ROADMAP.md` (§4 Math Correctness Charter, decision D7, and Phase 0 item 0.11). This is a faithful, high-quality implementation of the Charter — it does what 0.11 set out to do, and the commit-per-subsection structure mirrors the roadmap's own breakdown. I verified the claims locally rather than taking the PR description at its word.

### Charter §4 coverage — all five layers present and substantive
- **Layer 1 (reference/oracle)** — `math-reference.test.ts`: PMT/IPMT/CUMIPMT/FV with cited spreadsheet sources *and* independent closed-form cross-checks, plus a fully hand-derived 3-month schedule. Meets the "≥2 independent reference points per formula, source cited" bar.
- **Layer 2 (cross-implementation)** — `forecast-consistency.test.ts`: `forecastLoan` vs `generateAmortizationSchedule` to the cent, `forecastInvestment` vs `generateInvestmentGrowth` at compounding boundaries. Exactly the §4 layer-2 guarantee. Comparing in integer cents is the right way to state "agrees to the cent."
- **Layer 3 (property/invariant)** — `math-properties.test.ts`: fast-check covers money conservation in cents, non-negative balances, extra-payment/contribution monotonicity, and pointwise net-worth additivity — the §4 invariant list.
- **Layer 4 (edge cases)** — `math-edge-cases.test.ts`: leap-day anniversaries (incl. the 2100 century rule), month-end rollover, 0%/30% rates, one-month term, negative amortization, horizon=today / mid-month / 50-year. Matches the §4 catalog item-for-item.
- **Layer 5 (precision)** — `PRECISION.md` is the single documented rounding/consistency policy §4 asked for, and the suite honors it (exact `toBe` where pinned, justified `toBeCloseTo`).

### D7 boundary
The ESLint `no-restricted-imports` rules correctly scope `src/helpers/**` + `src/models/**` to a framework-free core (no react/react-dom/@mui/*/@emotion/*/react-number-format/`*.tsx`), turning D7's "purity is a convention" into a build failure as the roadmap specified. Relocating the `useFieldTracking` hook to `src/hooks/` is the right fix for the one offender, and the copilot-instructions update documents the new layout. This is the mechanical-extraction property D7 wants for the future `@pathwise/engine` graduation.

### Enforcement
Coverage gate wired into CI via `test:coverage` (and it goes beyond §4's stated line+branch to also pin functions+statements — stricter, good). `coverage.include: ['src/helpers/**/*.ts']` is correct, and Vitest's `coverage.all` defaults to `true`, so an untested helper file can't silently satisfy the gate. Mutation testing is correctly kept off the per-PR path (weekly + dispatch).

### Things to address (see inline)
1. **PRECISION.md step-up divergence** — the one real correctness gap. It's handled honestly, but it currently lives only as prose + a missing assertion, so the suite is green and silent about a known bug. The Charter's process rule wants a failing regression test (or `it.fails`/`it.todo`) and ideally a tracking issue. *(highest-value item)*
2. **stryker `break: 70`** — fine as a floor, but a non-blocking weekly run with a 70% threshold has no mechanism that forces surviving mutants toward the §4 "triage to zero" target. Suggest recording a baseline score and committing to ratchet.
3. **Month-end rollover edge test** — correctly pins the raw-`Date` quirk; worth a tracking issue so the Phase 6 dayjs migration revisits it deliberately.

### Verdict
Phase 0's correctness gate, landed well — this is the kind of foundation §4 was written to demand before Phase 2 puts numbers on charts. The inline notes are follow-ups/tracking, not blockers; the only one I'd genuinely want resolved before this is treated as "Phase 0 complete" is making the step-up divergence an executable test rather than a paragraph. Nice work.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by bberardi** [comment]
Thanks for the thorough review. Addressed in `e84e8a5` and `c76204e`:

**1. Step-up divergence → executable test (the actionable one).** Added an `it.fails` tripwire to `forecast-consistency.test.ts` asserting the two engines agree with a yearly step-up. It currently throws (they diverge at the first anniversary), so `it.fails` passes and the divergence is now visible in the runner (`1 expected fail`) instead of living only in prose. The day the off-by-one is reconciled, the body stops throwing and the test flips red — forcing whoever fixes it to convert it to a normal `it`, which is the Charter's "failing test before the fix." `PRECISION.md` §4 now points at that test.

**2. Stryker baseline + ratchet.** Ran the first full mutation pass: **85.33% overall** on `src/helpers/**` (844 killed / 147 survived / 11 timeout, 0 no-coverage). Recorded that baseline (with per-file scores) in `stryker.config.json` and bumped `break` 70 → 80, so the weekly run now fails on a mutation-score regression rather than being inert. Driving survivors toward zero is open-ended triage.

**3. Month-end rollover & the `toEqual(new Date(2021, 2, 3))` over-spec.** Agreed these are deliberate tight pins of the raw-`Date` quirk; left as-is so the Phase 6 dayjs migration has to change them consciously.

**4. `investment-helpers` dead-branch removal.** Thanks for confirming behavior-preserving.

Per maintainer direction, the remaining open-ended follow-ups (dayjs/month-end migration, mutation-survivor triage, and reconciling the step-up divergence itself) are being **deferred to the proposed ROADMAP §8.1 correctness backlog (#60)** rather than tracked as separate issues — so they live with the rest of that list. No `ROADMAP.md` change here to avoid colliding with #60.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by bberardi** [comment]
Update on tracking: opened dedicated issues for the non-blocking follow-ups so they're durable (they can still feed the §8.1 correctness backlog in #60):

- #62 — Reconcile the step-up anniversary off-by-one (the `it.fails` tripwire here flips red when this lands)
- #63 — Phase 6: migrate date math to dayjs / revisit the raw-`Date` month-end rollover pin
- #64 — Ratchet the Stryker mutation threshold toward zero survivors (baseline 85.33% recorded)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/helpers/PRECISION.md (hunk 1 above)
Comment (bberardi) [reply src/helpers/PRECISION.md]: Done in `e84e8a5`: added the `it.fails` tripwire to `forecast-consistency.test.ts` (visible in the runner now, flips red when the off-by-one is fixed), and `PRECISION.md` §4 points at it. Reconciling the divergence itself is tracked in #62. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: stryker.config.json (hunk 2 above)
Comment (bberardi) [reply stryker.config.json]: Done in `c76204e`: recorded the first full mutation pass (85.33% overall, per-file scores) in `stryker.config.json` and bumped `break` 70 → 80, so the weekly run now fails on a score regression instead of being inert. Survivor triage / further ratcheting is tracked in #64. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/helpers/math-edge-cases.test.ts (hunk 3 above)
Comment (bberardi) [reply src/helpers/math-edge-cases.test.ts:73]: Agreed — both the month-end pin and the `toEqual(new Date(2021, 2, 3))` day over-spec are deliberate tight pins of the raw-`Date` quirk, left as-is so the Phase 6 dayjs migration has to change them consciously. That migration (and revisiting this quirk) is tracked in #63. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_
